torch/nn/modules/module.py

```diff
@@ -729,14 +729,47 @@ def share_memory(self):
     def _get_name(self):
         return self.__class__.__name__
 
-    def __repr__(self):
-        tmpstr = self._get_name() + '(\n'
+    def get_extra_repr(self):
+        r"""Get the extra representation of the module
+
+        To print customized extra information, you should reimplement
+        this method in your own modules. Each item in the returned list will
+        be printed in separate line.
+        """
+        return []
+
+    def _get_extra_lines(self):
+        r"""Helper method to process extra representation"""
+        extra_repr = self.get_extra_repr()
+        if isinstance(extra_repr, str):
+            return extra_repr.split('\n')
+        else:
+            return extra_repr
+
+    def _get_child_lines(self):
+        child_lines = []
         for key, module in self._modules.items():
-            modstr = module.__repr__()
-            modstr = _addindent(modstr, 2)
-            tmpstr = tmpstr + '  (' + key + '): ' + modstr + '\n'
-        tmpstr = tmpstr + ')'
-        return tmpstr
+            mod_str = module.__repr__()
+            mod_str = _addindent(mod_str, 2)
+            child_lines.append('(' + key + '): ' + mod_str)
+        return child_lines
+
+    def __repr__(self):
+        main_str = self._get_name() + '('
+        # We treat the extra repr like the sub-module, one item per line
+        # TODO:the api is to be discussed later
+        extra_lines = self._get_extra_lines()
+        child_lines = self._get_child_lines()
+        lines = extra_lines + child_lines
+        if lines:
+            # simple one-liner info, which most builtin Moudules will use
+            if len(extra_lines) == 1 and not child_lines:
+                main_str += extra_lines[0]
+            else:
+                main_str += '\n' + ''.join(['  ' + line + '\n' for line in lines])
```

Comment:
Or better:
```python
'\n' + '\n'.join(_addident(line, 2) for line in lines)
```

Reply:
It's not going to work.
1) the `_addindent` does nothing when input string has only one line
2) the trailing new-line is missing.
Actually I come up with a concrete solution, but harder to read:
```
main_str += '\n  ' + '\n  '.join(lines) + '\n'
```